src/NLog/StructuredEvents/TemplateRenderer.cs

```diff
@@ -0,0 +1,149 @@
+                    var hole = holeEnumerator.Current.Hole;
+                    if (hole.Index != -1)
+                    {
+                        RenderHole(sb, hole, formatProvider, parameters[hole.Index], true);
+                    }
+                    else
+                    {
+                        RenderHole(sb, hole, formatProvider, parameters[holeIndex++]);
+                    }
+                }
+            }
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Render a template to a string.
+        /// </summary>
+        /// <param name="template">The template.</param>
+        /// <param name="formatProvider">Culture.</param>
+        /// <param name="parameters">Parameters for the holes.</param>
+        /// <returns>Rendered template, never null.</returns>
+        public static string Render(this Template template, IFormatProvider formatProvider, object[] parameters)
+        {
+            var sb = new StringBuilder(template.Value.Length + 64 * template.Holes.Length);
+            int pos = 0;
+            int holeIndex = 0;
+            foreach (var literal in template.Literals)
+            {
+                sb.Append(template.Value, pos, literal.Print);
+                pos += literal.Print;
+                if (literal.Skip == 0)
+                {
+                    pos++;
+                }
+                else
+                {
+                    pos += literal.Skip;
+                    if (template.IsPositional)
+                    {
+                        Hole hole = template.Holes[holeIndex++];
+                        RenderHole(sb, hole, formatProvider, parameters[hole.Index], true);
+                    }
+                    else
+                    {
+                        RenderHole(sb, template.Holes[holeIndex], formatProvider, parameters[holeIndex++]);
+                    }
+                }
+            }
+            return sb.ToString();
+        }
+
+        private static void RenderHole(StringBuilder sb, Hole hole, IFormatProvider formatProvider, object value, bool legacy = false)
+        {
+            RenderHole(sb, hole.CaptureType, hole.Format, formatProvider, value, legacy);
+        }
+
+        public static void RenderHole(StringBuilder sb, CaptureType captureType, string holeFormat, IFormatProvider formatProvider, object value, bool legacy = false)
+        {
+            if (value == null)
+            {
+                sb.Append("NULL");
+                return;
+            }
+
+            switch (captureType)
+            {
+                case CaptureType.Stringify:
+                    // TODO: we don't need to support format and alignment here?
+                    sb.Append('"').Append(value.ToString()).Append('"');
+                    break;
+                case CaptureType.Serialize:
+                    Config.ConfigurationItemFactory.Default.JsonConverter.SerializeObject(value, sb);
```

Comment:
I think we should split this.

e.g. we like to use semi-json for message templates, and json.net for real json in JsonLayout. What do you think? 

Maybe adding MessageTemplateSerializer property on ConfigurationItemFactory?